src/test/groovy/graphql/schema/idl/TypeDefinitionRegistryTest.groovy

```diff
@@ -397,24 +417,36 @@ class TypeDefinitionRegistryTest extends Specification {
 
             interface Mammal {
               id: String!
+              mother: Mammal!
+              offspring: [Mammal!]!
             }
 
             interface Reptile {
               id: String!
             }
 
-            type Dog implements Animal & Mammal {
+            interface Canine implements Animal & Mammal {
               id: String!
+              mother: Canine!
+              offspring: [Canine!]!
```

Comment:
Prior to this change, this schema would result in the following errors:
```
The interface type 'Canine' [@11:1] has tried to redefine field 'mother' defined via interface 'Mammal' [@5:1] from 'Mammal!' to 'Canine!'
The interface type 'Canine' [@11:1] has tried to redefine field 'offspring' defined via interface 'Mammal' [@5:1] from '[Mammal!]!' to '[Canine!]!'
```

Reply:
Hmmm I am not sure if this is allow in graphql. We will need to investigate this more in spec terms.

That is the field `mother : Mammal!` becomes a ` mother : Canine!` and while Cannie is a Mammal I am not sure if the graphql spec allows this type inference

Reply:
It's allowed by [this part of the spec](https://github.com/graphql/graphql-spec/blob/main/spec/Section%203%20--%20Type%20System.md#objects):

```
IsValidImplementationFieldType(fieldType, implementedFieldType):
  ...
  5. If {fieldType} is an Object or Interface type and
     {implementedFieldType} is an Interface type and {fieldType}
     declares it implements {implementedFieldType} then return {true}.
  ...
```

In this case, for the field `mother`, after unwrapping non-null, `{fieldType}` is `Canine`, and `{implementedFieldType}` is `Mammal`, and `Canine` implements `Mother`, so the condition is satisfied.

The reference implementation allows it as well:
- https://github.com/graphql/graphql-js/blob/6a5f51f3bd07f3914c7f1ec01908e9e325b77d58/src/type/validate.ts#L373:L382
- https://github.com/graphql/graphql-js/blob/6a5f51f3bd07f3914c7f1ec01908e9e325b77d58/src/utilities/typeComparators.ts#L72:L79